src/System.Security.Cryptography.Csp/src/System/Security/Cryptography/CspKeyContainerInfo.cs

```diff
@@ -238,15 +188,64 @@ public string UniqueKeyContainerName
         {
             get
             {
-                SafeProvHandle safeProvHandle = SafeProvHandle.InvalidHandle;
-                int hr = CapiHelper.OpenCSP(_parameters, (uint)CapiHelper.CryptAcquireContextFlags.CRYPT_SILENT, ref safeProvHandle);
+                return (string)ReadKeyParameterSilent(Constants.CLR_UNIQUE_CONTAINER);
+            }
+        }
+
+        /// <summary>
+        /// Read a the parameter from the current key using CRYPT_SILENT, to avoid any potential UI prompts.
+        /// </summary>
+        private object ReadKeyParameterSilent(int keyParam, bool throwOnNotFound=true)
+        {
+            const uint SilentFlags = (uint)CapiHelper.CryptAcquireContextFlags.CRYPT_SILENT;
+
+            SafeProvHandle safeProvHandle;
+            int hr = CapiHelper.OpenCSP(_parameters, SilentFlags, out safeProvHandle);
```

Comment:
According to MSDN (https://msdn.microsoft.com/en-us/library/windows/desktop/aa379886%28v=vs.85%29.aspx)

> For hardware-based CSPs, such as a smart card CSP, if the pszContainer parameter is NULL or blank, this flag implies that no access to any keys is required, and that no UI should be presented to the user. This form is used to connect to the CSP to query its capabilities but not to actually use its keys.

So it seems like that's the way to talk to the provider to ask it if it's hardware.